relay-redis/src/real.rs

```diff
@@ -34,28 +39,56 @@ enum ConnectionInner<'a> {
     Single(&'a mut redis::Connection),
 }
 
-impl ConnectionLike for ConnectionInner<'_> {
-    fn req_packed_command(&mut self, cmd: &[u8]) -> redis::RedisResult<redis::Value> {
+impl ConnectionInner<'_> {
+    fn get_connections(&mut self) -> Connections<'_> {
         match self {
-            ConnectionInner::Cluster(ref mut con) => con.req_packed_command(cmd),
+            ConnectionInner::Cluster(con) => Connections {
+                primary: con as &mut (dyn ConnectionLike + Send),
+                secondaries: vec![],
+            },
             ConnectionInner::MultiWrite {
                 primary: primary_connection,
                 secondaries: secondary_connections,
             } => {
-                let primary_result = primary_connection.req_packed_command(cmd);
+                let mut connections = primary_connection.get_connections();
+
                 for secondary_connection in secondary_connections.iter_mut() {
-                    if let Err(error) = secondary_connection.req_packed_command(cmd) {
+                    let secondary_connections = secondary_connection.get_connections();
+                    connections.secondaries.push(secondary_connections.primary);
+                    connections
+                        .secondaries
+                        .extend(secondary_connections.secondaries);
+                }
+
+                connections
+            }
+            ConnectionInner::Single(con) => Connections {
+                primary: con as &mut (dyn ConnectionLike + Send),
+                secondaries: vec![],
+            },
+        }
+    }
+}
+
+impl ConnectionLike for ConnectionInner<'_> {
+    fn req_packed_command(&mut self, cmd: &[u8]) -> redis::RedisResult<redis::Value> {
```

Comment:
All of this can be written much shorter and with allocations/without dynamic dispatch:

```rs
enum ConnectionInner<'a> {
    Cluster(&'a mut redis::cluster::ClusterConnection),
    MultiWrite {
        primary: Box<ConnectionInner<'a>>,
        secondaries: Vec<ConnectionInner<'a>>,
    },
    Single(&'a mut redis::Connection),
}

fn log_secondary<T>(result: redis::RedisResult<T>) {
    if let Err(error) = result {
        relay_log::error!(
            error = &error as &dyn Error,
            "sending cmds to the secondary Redis instance failed",
        );
    }
}

impl ConnectionLike for ConnectionInner<'_> {
    fn req_packed_command(&mut self, cmd: &[u8]) -> redis::RedisResult<redis::Value> {
        match self {
            ConnectionInner::Cluster(con) => con.req_packed_command(cmd),
            ConnectionInner::Single(con) => con.req_packed_command(cmd),
            ConnectionInner::MultiWrite {
                primary,
                secondaries,
            } => thread::scope(|s| {
                for connection in secondaries {
                    s.spawn(move || log_secondary(connection.req_packed_command(cmd)));
                }
                primary.req_packed_command(cmd)
            }),
        }
    }

    fn req_packed_commands(
        &mut self,
        cmd: &[u8],
        offset: usize,
        count: usize,
    ) -> redis::RedisResult<Vec<redis::Value>> {
        match self {
            ConnectionInner::Cluster(con) => con.req_packed_commands(cmd, offset, count),
            ConnectionInner::Single(con) => con.req_packed_commands(cmd, offset, count),
            ConnectionInner::MultiWrite {
                primary,
                secondaries,
            } => thread::scope(|s| {
                for connection in secondaries {
                    s.spawn(move || {
                        log_secondary(connection.req_packed_commands(cmd, offset, count))
                    });
                }
                primary.req_packed_commands(cmd, offset, count)
            }),
        }
    }
```

Reply:
Yeah without the generalization it becomes much simpler, good tip!